### ISSUE TITLE
Issue on inputs.win_perf_counters with scaling behaviour

### ISSUE THREAD

**Issue by thaentzler** [influxdata/telegraf#6448]
Hello, i encounter issue on win_perfcounter. 

   Seems like the value retrieve from process perf counter is calculated based on automated Scale (so we can retriev process % up to 500000 and another with 5 cause different scale is apply on each process name .....) . 

   I have check on pdh.go and seems like you do not apply the PDH_FMT_NOSCALE flag when value is retrieve. 

do you think it's possible to indicated this behaviour in plugins ?

**Comment by vlastahajek** [comment]
You mean you are getting values over 500.000 % for process CPU usage?

Scaling factor is provided by each counter to properly calculate displayable value by retrieving function for formatted counter values, by default.  It should be same for all instances for same counter.

Could you, please, provide your win_perf_counters plugin configuration and OS and HW info?

Could you, also, try `typeperf` util? e.g. `typeperf "\Process(*)\% Processor Time"` or so, please.

**Comment by thaentzler** [comment]
Issue here, is that we monitor a some windows Server (2012, 2008 .... etc) and we saw that depending on server, the scaling change dynamicaly. that's quite troublesome to compare consumption of a process between two server. 


For example on 2 server with same process : 

At 08H00AM 

Server 1 : 50.000,00 % for test.exe
Server 2 : 50,00% for test.exe

At 06H00PM 

Server 1 : 5.000,00 % for test.exe
Server 2 : 500,00% for test.exe

So scalling change on server, but on timeslot too ....

So if i want to make a representation in a single graph in grafana .... it's kind of complicated. 

I know that by setting the PDH_FMT_NOSCALE flag you can force the scaling to be at 1 for every counter.

**Comment by vlastahajek** [comment]
I definitely see your point for using PDH_FMT_NOSCALE. 

But it is still surprising to see such high values.  Scaling factor is usually provided to scale value to easily displayable range, which seems not to  be this case.

Could you, please, add this counter(s) to Perfmon and put its screenshot here? Perfmon shows scaling factors in the bottom.

It may happen that the counters got broken. 
You can  try to rebuild them:
1. open admin command line:
2. run `lodctr /R`. 
3. Check counter values again

**Comment by mbost** [comment]
Hello,

I'm a colleague of thaentzler (he will be out of office for a couple of week). We are facing this behavior on all Windows Servers on which we tested it (from serv 2008 to serv 2016).

I will give you an example where that scaling factor is really disturbing when we try to analyze server processes (I did a lodctr /R command before that analysis).
Below you have the perfmon Process CPU consumption of one of our server. I focused on 2 processes only :

![image](https://user-images.githubusercontent.com/56109057/66110400-fd737b80-e5c6-11e9-9035-14571817121f.png)

As you can see, the process diahost consumes around 30-40% of CPU and WmiPrvSE take around 2% which is nothing.
Now when we look at Telegraf on the same time window, here is when I can see (don't look at right Y axis) :

![image](https://user-images.githubusercontent.com/56109057/66110628-7672d300-e5c7-11e9-90b1-52da66a60993.png)

As you can see, process WmiPrvSE is showing around 20000% CPU consumption which means it might have a scale of 10000. But this is an issue for us because we don't have the visibility of process diahost. I guess this process has a scale of 1 in Telegraf and if we only select this process, we can see values we had on perfmon (see below) :

![image](https://user-images.githubusercontent.com/56109057/66110819-da959700-e5c7-11e9-96aa-db96d383ee16.png)

This behavior of scale factor interpretation is trending to hide processes that are really taking CPU ressources because their scale is lower so we can't see them on the graph.
The best for us would be to have a scale of 1 for every process so we could have the visibility on processes that are really taking the CPU.

Thank you for your help.

**Comment by vlastahajek** [comment]
Thanks @mbost for the details. It is really weird. 
In order to shed more light on this, could you, please, give me outputs of actual values from telegraf? 
`telegraf.exe -config <your_config_path> -test`. Feel free to remove parts of the output that is not relevant (other than process counter).

And it would be also very helpful,if you could give me few lines of output from: `typeperf "\Process(WmiPrvSE)\% Processor Time"`

**Comment by mbost** [comment]
Below is the output of --test command for both processes:

> win_proc,host=XXXXXX,instance=diahost,objectname=Process Handle_Count=1472,IO_Read_Bytes_persec=2389.1064453125,IO_Write_Bytes_persec=2923.641845703125,Page_Faults_persec=34.58177185058594,Percent_Processor_Time=0,Private_Bytes=145674240,Virtual_Bytes=865595392,Working_Set=214175744 1570103157000000000
> win_proc,host=XXXXXX,instance=WmiPrvSE,objectname=Process Handle_Count=391,IO_Read_Bytes_persec=0,IO_Write_Bytes_persec=0,Page_Faults_persec=49.402530670166016,Percent_Processor_Time=0,Private_Bytes=26750976,Virtual_Bytes=2199161667584,Working_Set=37842944 1570103157000000000

And below is the output of typeperf command for both processes:

![Capture1](https://user-images.githubusercontent.com/56109057/66126576-104b7780-e5ea-11e9-8f61-38484945e293.PNG)

**Comment by vlastahajek** [comment]
Thanks a lot. 
In the time you gathered this data, was your dashboard showing also such high values, as you shown above? 
If not, could you, please, reproduce that situation and gather this output again?
 
The reason I'm still asking for such data is that I'm trying to understand the performance counters behavior, cause it is a bit different on your machines.

**Comment by mbost** [comment]
Yes the dashboard was still showing this kind of data. Below is the graph for WmiPrvSE:

![image](https://user-images.githubusercontent.com/56109057/66133991-e1d49900-e5f7-11e9-9308-02c6b262c2ab.png)

And below il the graph with only diahost selected:

![image](https://user-images.githubusercontent.com/56109057/66134060-fd3fa400-e5f7-11e9-8984-a826b2490195.png)

**Comment by vlastahajek** [comment]
It's a bit pity that the typeperf output shows values from different time windows than it is shown on dashboard.
Would be possible to see output from PerfMon, typeperf and Dashboard during the same time period when a process shows high values in the dashboard?
Also, how many total cores has the server?

**Comment by mbost** [comment]
I have 4 cores on this server.
Below is the capture of the perfmon plus the typeperf processed simultaneously:

![Capture2](https://user-images.githubusercontent.com/56109057/66139352-92df3180-e600-11e9-98ee-ffbef613bec2.PNG)

Below is the output I have on my dashboard for WimPrvSE process:

![image](https://user-images.githubusercontent.com/56109057/66139445-bc985880-e600-11e9-9995-b23d09eecac9.png)

And below is the output on the dashboard for diahost process:

![image](https://user-images.githubusercontent.com/56109057/66139501-db96ea80-e600-11e9-9b74-d197a0ddfca3.png)

**Comment by vlastahajek** [comment]
Thank you very much  for your assistance, @mbost.
But I still have to bug you, sorry. 
I see a peak of CPU usage by WmiPrvSE on  the dashboard, but not values for the same time frame in perfmon or in typeperf output. 
Could you, please, add that?

**Comment by mbost** [comment]
Hello,

I know that you need to make sure that I am on the right server but trust me I am. It is really hard to make a short and valuable capture to compair Telegraf and typeperf because Telegraf is sampling 1 point every 15 seconds and I believe the point will not be printed on the graph at the exact time he is printed on typeperf.

The most important thing to know is that both processes are taken on the same server : MPLWIN0536. You can see on the below capture that both processes are taken on this server and in the legend we has max, min and avg values of WmiPrvSE and diahost process:

![image](https://user-images.githubusercontent.com/56109057/66187933-3fafc200-e686-11e9-89e5-ae796f0b18a5.png)

Please have a look at the legend, you can see WmiSrvPE has 131211% max value and 78683% average and at the same time diahost has 69% max value and 32% average.
You can also see at the top of the screen that the server selected is the right one. Please don't pay attention to the scale parameter at the top of the page, this was a test we did but it didn't work has entended.

**Comment by vlastahajek** [comment]
Friend, I'm very sorry if I'm confusing you. I'm not worried about you are being on wrong server or providing invalid values. I'm unable to reproduce and never seemed before. this, to I'm asking for a debug data.
 
I would just like to see the raw output from counters (typeperf, Prefmon) when your dashboard shows huge values. View on a dashboard is for me just interpreted  vision.

If you could, please, just leave typeperf reading on background, same with perfmon, monitoring WmiProvSE and reproduce highload that would be what all I need.

**Comment by vlastahajek** [comment]
I've just noticed WmiPrvSE can have multiple instances. 

Are you using wildcards expansion in Telegraf config?

When running typeperf, use this: `typeperf "\Process(WmiPrvS*)\% Processor Time"`, please.

**Comment by mbost** [comment]
Below is the stanza in my telegraf.conf for processes:

![image](https://user-images.githubusercontent.com/56109057/66208481-77822e00-e6b5-11e9-8348-2006a1dff538.png)

Also below is the configuration of the Telegraf dashboard:

![image](https://user-images.githubusercontent.com/56109057/66208554-97b1ed00-e6b5-11e9-921d-c2e8fd6223ca.png)

I have performed the test again using this time the wildcard but values are still really strange:

Typeperf + Permon: ![image](https://user-images.githubusercontent.com/56109057/66208626-d6e03e00-e6b5-11e9-9324-de568ca3c103.png)

Dashboard: ![image](https://user-images.githubusercontent.com/56109057/66208605-c039e700-e6b5-11e9-863e-1da4ff45bafa.png)

Thank you for your help

**Comment by vlastahajek** [comment]
Interesting, thanks a lot. 
Please, show also global part of  config: [[input.win_perf_counters]]

Would it be possible to see what is exactly written in this metric in the database for WmiPrvSE in the time window as used above ( between 14:37:50 and 14:38:55) ?

**Comment by mbost** [comment]
Below is the configuration for [[input.win_perf_counters]]

[[inputs.win_perf_counters]]
  [[inputs.win_perf_counters.object]]
    # Processor usage, alternative to native, reports on a per core.
    ObjectName = "Processor"
    Instances = ["*"]
    Counters = [
      "% Idle Time",
      "% Interrupt Time",
      "% Privileged Time",
      "% User Time",
      "% Processor Time",
      "% DPC Time",
    ]
    Measurement = "win_cpu"
    # Set to true to include _Total instance when querying for all (*).
    IncludeTotal=true

  [[inputs.win_perf_counters.object]]
    # Disk times and queues
    ObjectName = "LogicalDisk"
    Instances = ["*"]
    Counters = [
      "% Idle Time",
      "% Disk Time",
      "% Disk Read Time",
      "% Disk Write Time",
      "Current Disk Queue Length",
      "% Free Space",
      "Free Megabytes",
    ]
    Measurement = "win_disk"
    # Set to true to include _Total instance when querying for all (*).
    #IncludeTotal=false

  [[inputs.win_perf_counters.object]]
    ObjectName = "PhysicalDisk"
    Instances = ["*"]
    Counters = [
      "Disk Read Bytes/sec",
      "Disk Write Bytes/sec",
      "Current Disk Queue Length",
      "Disk Reads/sec",
      "Disk Writes/sec",
      "% Disk Time",
      "% Disk Read Time",
      "% Disk Write Time",
    ]
    Measurement = "win_diskio"

  [[inputs.win_perf_counters.object]]
    ObjectName = "Network Interface"
    Instances = ["*"]
    Counters = [
      "Bytes Received/sec",
      "Bytes Sent/sec",
      "Packets Received/sec",
      "Packets Sent/sec",
      "Packets Received Discarded",
      "Packets Outbound Discarded",
      "Packets Received Errors",
      "Packets Outbound Errors",
    ]
    Measurement = "win_net"

  [[inputs.win_perf_counters.object]]
    ObjectName = "System"
    Counters = [
      "Context Switches/sec",
      "System Calls/sec",
      "Processor Queue Length",
      "System Up Time",
    ]
    Instances = ["------"]
    Measurement = "win_system"
    # Set to true to include _Total instance when querying for all (*).
    #IncludeTotal=false

  [[inputs.win_perf_counters.object]]
    # Example query where the Instance portion must be removed to get data back,
    # such as from the Memory object.
    ObjectName = "Memory"
    Counters = [
      "Available Bytes",
      "Cache Faults/sec",
      "Demand Zero Faults/sec",
      "Page Faults/sec",
      "Pages/sec",
      "Transition Faults/sec",
      "Pool Nonpaged Bytes",
      "Pool Paged Bytes",
      "Standby Cache Reserve Bytes",
      "Standby Cache Normal Priority Bytes",
      "Standby Cache Core Bytes",

    ]
    # Use 6 x - to remove the Instance bit from the query.
    Instances = ["------"]
    Measurement = "win_mem"
    # Set to true to include _Total instance when querying for all (*).
    #IncludeTotal=false

  [[inputs.win_perf_counters.object]]
    # Example query where the Instance portion must be removed to get data back,
    # such as from the Paging File object.
    ObjectName = "Paging File"
    Counters = [
      "% Usage",
    ]
    Instances = ["_Total"]
    Measurement = "win_swap"

  [[inputs.win_perf_counters.object]]
    # Process metrics, in this case for IIS only
    ObjectName = "Process"
    Counters = ["% Processor Time","Handle Count","Private Bytes","Virtual Bytes","Working Set","IO Read Bytes/sec","IO Write Bytes/sec","Page Faults/sec"]
    Instances = ["*"]
    Measurement = "win_proc"
    #IncludeTotal=false #Set to true to include _Total instance when querying for all (*).

And below are the values of this process in the Prometheus database. It's GMT time that's why the the timerange is 2 hours before

![image](https://user-images.githubusercontent.com/56109057/66216667-2b8bb500-e6c6-11e9-8b1a-930d1f38bb4d.png)

**Comment by vlastahajek** [comment]
Thanks. 
We had to dig into more, cause it still seems to me, that it has nothing to do with scaling. 
Applied scaling means, that formatted values are in the range of 0-100. You can check some *Bytes counters, they have scaling like 0.0001 (-4), cause their values are really large.

To avoid further questions for config, could you, please add full config? Of course, you replace any sensitive info.

I see, you are using standard global counter configuration, which has wildcard expansion switched off. A major benefit of this feature is it will report values for multiple instances of a process, as you could see WmiPrvSE had 5 instances and you are monitoring only first. Major drawback is that it localizes counters on non-English systems.

Would it be possible to see a tabular view?

**Comment by mbost** [comment]
Hello,

Below is the zip of the telegraf.conf file.
[telegraf.zip](https://github.com/influxdata/telegraf/files/3695841/telegraf.zip)

We thought these values were due to Windows scaling because we know that Windows is aplying scales like 1000 for really low metrics. The WmiPrvSE is one of these low metrics so that sounded logical to have a scale increase on this one. On the other hand, process diahost is supposed to take all 4 CPU at a time so there is no scale factor on it as the scale is the same as "Total_" metric.

For the tabular view, I believe you need the capture below ?

![image](https://user-images.githubusercontent.com/56109057/66291048-18a0fc80-e8e1-11e9-846f-5d2dd0e5e92e.png)

Thank you

**Comment by mbost** [comment]
Hi,

For information, I've installed a Telegraf agent on Windows Servers 2008 and it looks like we don't have this issue. Every process have the %value as intended. Tell me if you need more information

**Comment by vlastahajek** [comment]
Thanks for new information. Yeah, the scale factor is somehow misleading.  
As you saw in conducted tests, typeperf, which reads counters the same way as telegraf, reports gathered values in ones of %, but final values in db are much larger.
So, it seems the issue should be somewhere in processing path.

Interesting finding with new installation. You mentioned [in a comment earlier](https://github.com/influxdata/telegraf/issues/6448#issuecomment-537846442) you are experiencing this issue on multiple servers 
Is there ant difference between the new installation, without the issue, and the rest? (hw config, sw versions)

**Comment by mbost** [comment]
At the moment Telegraf is installed on 2008, 2012 and 2016 servers and I was wrong, the issue is only for 2012 and 2016. Sorry for that mistake.
But this the same version and the same configuration file for each server.

**Comment by vlastahajek** [comment]
Ok.  And the reproduction we did above was done what server on?

**Comment by mbost** [comment]
All the captures above were taken for a Windows Server 2016 in 64 bits.
Yes all the servers with the issue are 64 bits and the 2008 are also 64 bits.

**Comment by vlastahajek** [comment]
Could, you, please, add the _file output_ to the telegraf config, run the suite shortly where you will see high numbers in dashboard and attach here the output? 
```
[[outputs.file]]
  files = ["c:/tmp/telegraf-output.txt"]
```
What version of Telegraf do you run?

**Comment by mbost** [comment]
Hello,

Please find attached this logfile for all the day of the 10th October starting at 8AM.

[telegraf_101018.log](https://github.com/influxdata/telegraf/files/3715970/telegraf_101018.log)

Thanks

**Comment by vlastahajek** [comment]
Thanks a lot.  There other three processes reporting high values.

I know it is getting tedious, but I'm not able to reproduce it (so far tried Windows 10, Windows Server 2012 R2 and 2019 DC), so I need your assitance. I really appreciate your help.
 
What happens at the server, when you are getting those values?

Could you, please, make following change in the configuration, run the reproduction (what happens between 8:45 and 10:00 UTC) again and attach the output? It will use a another (but not much different) way how to gather counter values.
```
[[inputs.win_perf_counters]]
  UseWildcardsExpansion=true
  UsePerfCounterTime=true
  CountersRefreshInterval="10s"
```

**Comment by mbost** [comment]
Hello,

Thank you so much, I believe these were the parameters I needed for Telegraf to work as entended !
Below is the Process %CPU over past weekend for the same server after I added the new configuration :

![image](https://user-images.githubusercontent.com/56109057/66733320-72647200-ee5f-11e9-920d-3cc2cb5c3c92.png)

The server has 4 CPU and this consumption is the real one, the one I wanted to have. I guess these parameters were needed for higher releases than Windows Server 2008 R2.

Thank you again @vlastahajek !

**Comment by mbost** [comment]
I have a last question : What is the equivalent of "UseWildcardsExpansion" parameter for Linux system ?

**Comment by vlastahajek** [comment]
I'm really glad I could help you with resolving the issue.

The fact that using `UseWildcardsExpansion=true` solved your problem means there is still an issue when reading counters in standard mode, either in Telegraf or in Windows.
Would you be open if I gave you later a custom built debug application to gather some values using different methods and debug this more? 

The `UseWildcardsExpansion` param directs win_perf_counters plugin to expand wildcards found in given counter-path during configuration and then it regularly reads metric for each counter individually. 
This why it needs the  `CountersRefreshInterval` for specifying how often it should reread available counters. This is mostly important in case of multi instance counters where the set of instances varies frequently (e.g. processes). The major advantage is that in this mode it can detect instance indexes (same process running in more instances), which is not possible in the plugin's standard mode.
It may happen that you can notice regular CPU spikes (not detected by Telegraf), which are caused by the counters re-reading. I suggested you `CountersRefreshInterval="10s"` for first try, as it is the same as your agent refresh interval, but you can prolong it a bit (30s, 1m), depending on your environment.

Regarding linux, Telegraf has another plugin for gathering system metrics on linuxes and it works differently. Do you have any issue with this?

**Comment by mbost** [comment]
Thank you for all these explanations.

We don't really have an issue for linux systems as the values of CPU Process consumption are the good one, but we only have 1 instance of each process. I guess this is a parameter to add as you did for windows system with wildcards expansion.

Of course I am available to make any test you need with a custom version

**Comment by powersj** [comment]
Hi Folks,

Reading through old issues and seeing status. Per the closed PR, #7007, it seems that using the `UseWildcardsExpansion = true` option helped resolve this issue. I am going to close this issue due to this fix and lack of any further updates or comments, but if I have missed something, please do let me know.

Thanks!
